hoomd/md/bond.py

```diff
@@ -381,3 +381,86 @@ def set_from_file(self, bondname, filename):  # noqa
                             rmin=rmin_table,
                             rmax=rmax_table,
                             coeff=dict(V=V_table, F=F_table, width=self.width))
+
+
+class Tether(Bond):
+    r"""Tethering bond potential.
+
+    :py:class:`Tether` specifies a Tethering potential energy between two
+    particles in each defined bond.
+
+    The tethered network is described in Refs.
+    `Gompper, G. & Kroll, D. M. in Statistical Mechanics of Membranes and
+    Surfaces 2nd edn (eds Nelson, D. R. et al.) 359-426 (World Scientific, 2004)
+    <https://www.worldscientific.com/worldscibooks/10.1142/5473>`_ and `Noguchi
+    , H. & Gompper, G., Phys. Rev. E 72 011901 (2005)
+    <https://link.aps.org/doi/10.1103/PhysRevE.72.011901>`_.
+
+    .. math::
+
+        V(r) = V_{\mathrm{att}}(r) + V_{\mathrm{rep}}(r)
+
+    where :math:`\vec{r}` is the vector pointing from one particle to the other
+    in the bond.
+
+    .. math::
+        :nowrap:
+
+        \begin{eqnarray*}
+        V_{\mathrm{att}}(r)  = & k_b \frac{exp(1/(l_{c0}-r)}{l_{max}-r}
+                                & r > l_{c0}\\
+                                = & 0
+                                & r \leq l_{c0}
+        \end{eqnarray*}
+
+    .. math::
+
+        \begin{eqnarray*}
+        V_{\mathrm{rep}}(r)  = & k_b \frac{exp(1/(r-l_{c1})}{r-l_{min}}
+                                & r < l_{c1}\\
+                                = & 0
+                                & r \ge l_{c1}
+        \end{eqnarray*}
+
+    .. math::
+        l_{min} < l_{c1} < l_{c0} < l_{max}
```

Comment:
I think it is a good idea to raise an error if this requirement is not met.

Reply:
Sounds great. Could you let me know how I can work around TypeParameter before arguments are assigned to cpp?

Reply:
I would suggest to raise a `ValueError` by setting up a postprocess function. Something similar as in the `ReversePerturbationFlow` class in `update.py`.

Reply:
Since individual parameter in the dict cannot be validated unless validated as a whole, I added the ValueError in EvaluatorBondTether.h under pybind11 dict. I tested with an invalid set of parameters and got the ValueError with inserted warning message.
If this is not the most elegant way to throw an error and better be addressed in bond.py, I can make the change later in separate branch.

Reply:
If there's a better way to use if statement to check the order of params, please let me know!